### PR TITLE
feat(dssubs): implement explicit queue declare + provide APIs

### PR DESCRIPTION
**Pull request by keynslug**
Fixes [EMQX-12153](https://emqx.atlassian.net/browse/EMQX-12153).

Release version: e5.8.1

## Summary

See individual commits.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] **Schema changes are backward compatible**
    API Schema has changed slightly.

[EMQX-12153]: https://emqx.atlassian.net/browse/EMQX-12153?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_store.erl [hunk 1]
```diff
@@ -263,67 +284,87 @@ mk_leader_topic(GroupName) ->
...
+        properties => #{},
         seqnum => 0,
         seqmap => #{},
+        committed => 0,
         stage => #{}
     }.
 
-open_store(Store = #{group := Group}) ->
-    ReadRootset = mk_read_root_batch(Group),
+open_rootset(ID) ->
```
Comment (savonarola) [inline apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_store.erl:341]: If we implement synchronous version of `get_streams`, will this simplify to a simple read operation?

File: apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl [hunk 2]
```diff
@@ -238,159 +238,133 @@ do_subscriptions_query_mem(QString) ->
     end.
 
 do_subscriptions_query_persistent(#{<<"page">> := Page, <<"limit">> := Limit} = QString) ->
-    Count = emqx_persistent_session_ds_router:stats(n_routes),
-    %% TODO: filtering by client ID can be implemented more efficiently:
-    FilterTopic = maps:get(<<"topic">>, QString, '_'),
-    Stream0 = emqx_persistent_session_ds_router:stream(FilterTopic),
-
+    PageSize = 100,
```
Comment (savonarola) [inline apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl:241]: This is a bit confusing: seems this is "internal" iterator page size, while we are in a paginated request where page size is Limit

**Review by savonarola — APPROVED** [review]

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_leader.erl [hunk 3]
```diff
@@ -99,8 +95,9 @@
 start_link(Options) ->
     gen_statem:start_link(?MODULE, [Options], []).
 
-become(ShareTopicFilter, StartTime, Claim) ->
-    Data0 = init_data(ShareTopicFilter, StartTime),
+become(ShareTopicFilter, Claim) ->
+    Data0 = init_data(ShareTopicFilter),
+    Data0 =/= false orelse exit(shared_subscription_not_declared),
```
[the remark this answers is not in the document]
Comment (keynslug) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_leader.erl:100]: Yes. I guess we can crash in a more ugly way instead.

**Review by savonarola — DISMISSED** [review]

**Review by savonarola — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-4367)

**Pull request merged**